feat(plugin-image-gallery): Image Grid (PE-53, PE-54)

shn-srl opened the pull request:
**Linear Issues**:
- PE-54
- PE-53

**Demo Links**: 
- [Editor Preview](https://frontend-git-pe-54-plugin-image-gallery-serlo.vercel.app/___editor_preview?state={"plugin"%3A"rows"%2C"state"%3A[{"plugin"%3A"imageGallery"%2C"state"%3A{"images"%3A[{"plugin"%3A"image"%2C"state"%3A{"src"%3A"https%3A%2F%2Fwww.shutterstock.com%2Fimage-vector%2Ftwo-opponents-facing-each-other-260nw-2296619399.jpg"%2C"caption"%3A{"plugin"%3A"text"%2C"state"%3A[{"type"%3A"p"%2C"children"%3A[{"text"%3A""}]}]%2C"id"%3A"4f3872a4-0900-43fc-bb02-85c218320616"}}%2C"id"%3A"07ae3472-4b40-4a8c-8605-ee3efedabf69"}%2C{"plugin"%3A"image"%2C"state"%3A{"src"%3A"https%3A%2F%2Fc6.staticflickr.com%2F9%2F8520%2F28357073053_cafcb3da6f_b.jpg"%2C"caption"%3A{"plugin"%3A"text"%2C"state"%3A[{"type"%3A"p"%2C"children"%3A[{"text"%3A""}]}]%2C"id"%3A"a251f2ed-a184-45ff-b025-1d47267dbccd"}}%2C"id"%3A"d934acaa-6f8f-4701-934a-aeeb1daabd3c"}%2C{"plugin"%3A"image"%2C"state"%3A{"src"%3A"https%3A%2F%2Fc8.staticflickr.com%2F9%2F8104%2F28973555735_ae7c208970_b.jpg"%2C"caption"%3A{"plugin"%3A"text"%2C"state"%3A[{"type"%3A"p"%2C"children"%3A[{"text"%3A""}]}]%2C"id"%3A"6c47127b-d410-46ef-a7e3-a2692766dd33"}}%2C"id"%3A"16fd4402-9b52-4238-a174-d8b4c1180688"}%2C{"plugin"%3A"image"%2C"state"%3A{"src"%3A"https%3A%2F%2Fcdn.pixabay.com%2Fphoto%2F2024%2F02%2F15%2F14%2F31%2Fdonkey-8575524_960_720.jpg"%2C"caption"%3A{"plugin"%3A"text"%2C"state"%3A[{"type"%3A"p"%2C"children"%3A[{"text"%3A""}]}]%2C"id"%3A"f8612633-273a-4ae0-a2f4-04e4da3ecc4c"}}%2C"id"%3A"2a264899-87a6-46fd-84be-067c4bbc0291"}]%2C"orderedIds"%3A""}%2C"id"%3A"c1efe02e-9554-4122-ba26-7667234411a1"}]})
- [Create Article](https://frontend-git-pe-54-plugin-image-gallery-serlo.vercel.app/entity/create/Article/1377)

**To do:**

- [x] Bug: Uploading more than two images makes the plugin break
- [x] Loading state for image upload
- [x] Single image upload

shn-srl committed:
chore(plugin-image-gallery): install react-photo-album library

shn-srl committed:
feat(plugin-image-gallery): Add Image Gallery view to editor

shn-srl committed:
feat(plugin-image-gallery): Add Image Grid in static view

shn-srl committed:
feat(plugin-image-gallery): Restore editor view correctly based on state

shn-srl committed:
refactor(plugin-image-gallery): Combine helper files, cleanup

shn-srl committed:
refactor(plugin-image-gallery): Remove unused function

shn-srl committed:
feat(plugin-image-gallery): Open single photo in modal when user clicks on photo in grid

shn-srl committed:
fix(plugin-image-gallery): Set view correctly when closing SingleImageModal, based on whether an image was uploaded or set

elbotho commented:
Thanks for the demo Links, really helpful here. Some initial feedback:

**Lightbox**

- tall images are clipping for me 
- when opening an image it's not selected and no toolbar is visible (only after click)
- will there be a lightbox for the static view at some point?


<img width="545" alt="image" src="https://github.com/user-attachments/assets/7e92fdf8-fced-47db-950e-080f61d200ae">
<img width="538" alt="image" src="https://github.com/user-attachments/assets/4c009a49-687f-4ce4-8363-1418de34466b">

--- 

**Bundle size**

The bundle size increase comes from the gallery plugin I guess? I think especially for the static view this is a lot of js just for the layout. I'm aware that there will be sorting later and that can be tricky to implement.

<img width="733" alt="image" src="https://github.com/user-attachments/assets/fb034485-2313-4c36-86ab-a49553ade03c">

Personally I'd love a simple css based solution for the grid and then using some library or code based on react-dnd only inside the editing view for sorting. (e.g. like [react-easy-sort](https://github.com/ValentinH/react-easy-sort) might be worth a look). But I can also not say with confidence how much work this way would actually be.

At least we should load the component dynamically.

shn-srl commented:
> Personally I'd love a simple css based solution for the grid and then using some library or code based on react-dnd only inside the editing view for sorting. (e.g. like [react-easy-sort](https://github.com/ValentinH/react-easy-sort) might be worth a look). But I can also not say with confidence how much work this way would actually be.
> 
> At least we should load the component dynamically.

I have a branch with the sorting functionality done with react-dnd. Can look at a css only solution for the grid



> 
> **Lightbox**
> 
> 
>     * will there be a lightbox for the static view at some point?

This is in another issue: https://linear.app/serlo/issue/PE-57/lightbox-desktop

shn-srl commented:
> Thanks for the demo Links, really helpful here. Some initial feedback:
> 
> **Lightbox**
> 
>     * tall images are clipping for me
> 
>     * when opening an image it's not selected and no toolbar is visible (only after click)
> 
>     * will there be a lightbox for the static view at some point?

This is in another issue:

shn-srl committed:
feat(plugin-image-gallery): Create custom ImageGrid component

shn-srl committed:
chore(plugin-image-gallery): Remove react-photo-album library

shn-srl committed:
fix(plugin-image-gallery): Add space above modal to avoid clipping of tall images

shn-srl committed:
feat(plugin-image-gallery): Always show image toolbar for image plugin inside ImageGallery plugin

shn-srl committed:
refactor(plugin-image-gallery): Use id instead of key for GridImage type

shn-srl commented:
@elbotho I have fixed the clipping (as far as I can tell), and the toolbar visibility.

Removed the react-photo-album library and implemented a solution with tailwind and styles

shn-srl committed:
fix(plugin-image-gallery): Fix grid behaviour

elbotho committed:
fix: make sure images don't clip and image does not overflow out of the modal

elbotho commented on packages/editor/src/plugins/image-gallery/index.ts:
```diff
@@ -7,17 +7,15 @@ import {
-      })
-    ),
+    images: list(child({ plugin: EditorPluginType.Image })),
+    orderedIds: string(''),
```
do we need `orderedIds` in the editor state? 
I thought the (ordered) array would have this information already.

shn-srl replied:
Removed

elbotho reviewed (commented):
> @elbotho I have fixed the clipping (as far as I can tell), and the toolbar visibility.
> 
> Removed the react-photo-album library and implemented a solution with tailwind and styles

Thank you for the adaptations!
I added another clipping fix, it seems it only happened on smaller screens (vertical space).

Bundle size still goes crazy, but maybe that's because of the CI being weird. I build it locally and it seemed fine.

I like the simpler layout approach 👍
Something [like this](https://play.tailwindcss.com/tNkRUbIsum) would also work nicely, but also has more dom elements. Not sure.

shn-srl committed:
refactor(plugin-image-gallery): Remove orderedIds from Image Gallery plugin

shn-srl commented:
> > @elbotho I have fixed the clipping (as far as I can tell), and the toolbar visibility.
> > Removed the react-photo-album library and implemented a solution with tailwind and styles
> 
> Thank you for the adaptations! I added another clipping fix, it seems it only happened on smaller screens (vertical space).
> 
> Bundle size still goes crazy, but maybe that's because of the CI being weird. I build it locally and it seemed fine.
> 
> I like the simpler layout approach 👍 Something [like this](https://play.tailwindcss.com/tNkRUbIsum) would also work nicely, but also has more dom elements. Not sure.

![image](https://github.com/user-attachments/assets/5ef14f8c-1c7e-4ffa-92a6-03404b0317ba)

Nice, thanks for the change.

I see the bundles size reduced.

Removed the orderedIds

shn-srl committed:
fix(plugin-image-gallery): lint error

elbotho commented:
this is the most relevant line:
![image](https://github.com/user-attachments/assets/9c862fdf-85bb-4e7b-a821-bcf2f31f0430)

hejtful committed:
fix(image-gallery): focus fixes

hejtful committed:
fix(image-gallery): more than two images uploaded bug

hejtful committed:
fix(image-gallery): gap 16px without outer padding

hejtful committed:
refactor(image-gallery): extract row percentage width util

hejtful committed:
feat(image-gallery): loading state

hejtful committed:
fix(image-gallery): fix focus issue when closing modal

hejtful committed:
fix(image-gallery): call the multiple upload callback even if only one file uploaded

hejtful committed:
fix(image-gallery): unnecessary re-renders

hejtful committed:
fix(image-gallery): focus image gallery plugin after image upload

hejtful committed:
refactor(image-gallery): remove unused reject

hejtful committed:
refactor(image-gallery): shorten multiple upload callback name

hejtful committed:
refactor(image-gallery): simplify JSX

hejtful committed:
refactor(image-gallery): improve comment

hejtful committed:
fix(image-gallery): close modal on pixabay and url image selection

elbotho committed:
refactor(plugin-gallery): rename photo to image

elbotho committed:
refactor(plugin-gallery): simplify image grid classes, max-height for last (single) image

elbotho committed:
refactor(plugin-gallery): use onClick to fix right mouse behaviour, support keyboard nav

elbotho committed:
fix(plugin-gallery): add missing +1 to index

elbotho committed:
refactor(plugin-gallery): proper naming for props and handlers
